pages/[[...slug]]/index.tsx

```diff
@@ -0,0 +1,43 @@
+import getAllPageContentsWithSlugString from '../../utilities/getAllPageContentsWithSlugString'
+import getPageContentBySlugString from '../../utilities/getPageContentBySlugString'
+import type { GetStaticPaths, GetStaticProps } from 'next'
+import PageContent from '../../components/Main/PageContent'
+import type { PageContentProps } from '../../components/Main/PageContent'
+import type { PageContentsWithSlugString } from '../../utilities/getAllPageContentsWithSlugString'
+import React from 'react'
+
+type SitePageProps = {
+  pageContent: PageContentProps
+}
+
+export default function SitePage({ pageContent }: SitePageProps): JSX.Element {
+  return <PageContent content={pageContent.content} title={pageContent.title} />
+}
+
+export const getStaticProps: GetStaticProps = async ({ params }) => {
+  let slugString: string
+
+  if (params?.slug) {
+    // params.slug = [ 'demo', 'js' ]
```

Comment:
Small thing but just before this is ready to merge could we remove the commented code from this file (I'm assuming it is no longer needed) 😁

Reply:
Hi, that is an example of the data type. Could that be helpful? Maybe I should put it somewhere else?

Reply:
I will use types instead of comments.

Reply:
This response worries me. Types are not alternatives to comments. Types are required. They are a form of static testing. They ensure that your code is correct -- that you are passing and returning the correct data types. The TypeScript type checker should be running on them and it should not be possible to make a commit or merge to main if the types don't check.

Comments do none of the above. Well-written code should have zero comments. The only reason for a comment is to explain a workaround if it cannot be avoided -- and then only to prevent the next coder (who might be you) from undoing the workaround and then discovering why it was there. As soon as the workaround can be safely removes, so, too, should the comment.

So no comments. Everything should be typed, and nothing should have type "any".

But there is another problem here and I'm surprised that Reuben missed it. `let` is for mutability. Everything in this code should be immutable, so we should never be using `let` (and absolutely never `var`). Everything should be `const`, but that's not enough. `const` is not immutable, only un-reassignable. So we must be careful not to mutate objects or arrays assigned with `const`.

Reply:
Oh yeah I did miss the use of `let` and reassignment but yeah that code should be changed :grin:

`getStaticPaths` is a Next.js thing. For a static site this function generates a page for each path in the returned array.

Reply:
So I see. But what is Next.js using it for? Why does it need it? Is it for static site generation?

I'm guessing that's the purpose (I don't remember using it with SSR before). Makes sense that it would need to know what pages to generate. Hmmm.

It would be interesting to know whether Contentful can a) store actual string arrays, and b) use them as a primary (or secondary) key, so that we can use the slug array for look up. That would avoid casting back and forth. And if we have to return it as an array anyway, maybe `stringify` and `parse` are a better way to go.

Reply:
BTW, does this work for all pages (i.e., we don't need the `pages/index.tsx` file anymore? That would be cool.

Reply:
It works for all pages. We don't need the pages/index.tsx anymore.

Reply:
Nice.

Reply:
[getStaticPaths](https://nextjs.org/docs/basic-features/data-fetching#getstaticpaths-static-generation) is required for dynamic routes:
> If a page has dynamic routes (documentation) and uses getStaticProps it needs to define a list of paths that have to be rendered to HTML at build time.

So I think that means if we use `params.slug` in the getStaticProps, then it is mandatory to have this function.